### PR TITLE
RFC: Add ODROID-U2/U3 support

### PR DESCRIPTION
**Pull request by jerome-benoit**
It's the minimal changes needed in LE to add support for ODROID-U2/U3 properly.

It's a major cleanup of all code floating around with a bunch of bugs fixed.
I will of course maintain this portage and offer support if it's merged.

It's a preliminary work to support more recent hardkernel hardware cleanly
and more easily.

Once the RFC is reviewed, the patch will be broken into small changeset for merging.

EDIT:

Test images for SD card and eMMC module of LibreELEC for ODROID-U2/U3 are available here : http://people.piment-noir.org/~fraggle/download/libreelec/Odroid_U2/ 

Signed-off-by: Jérôme Benoit <jerome.benoit@piment-noir.org>

### CODE REVIEW

**Comment by lrusak** [comment]
I'm not sure if this is something we are ready/looking for, but others can chime in.

Here are some comments:

 We don't need all the linux patches, they don't need to be included in this PR.

Does it need xorg? can it not run on the fbdev or is there no hardware support?

**Comment by jerome-benoit** [comment]
Le 20/12/2016 à 08:05, Lukas Rusak a écrit :
>
> I'm not sure if this is something we are ready/looking for, but others
> can chime in.
>

I think you should introduce community hardware support or secondary
hardware support:
People doing the portage will benefit from LE infrastructure and
audience to ship their portage (update channel, addons channels, sources
tarball hosting, etc.).
LE is not responsible if something break in the secondary hardware
images shipped and is free to remove them if the maintainer is not
responsive anymore.
I think the best way to achieve this without polluting the LE master git
branch, it's to create a master-secondary or master-community branch
that will stay in sync with the master branch automatically (it's a
matter of git configuration) and generate tags from here to build and
ship community/secondary LE image for the hardware in sync with the main
LE schedule.

> Here are some comments:
>
> We don't need all the linux patches, they don't need to be included in
> this PR.
>

Every patches in here against the old kernel version shipped by
hardkernel for ODROID-U3 is needed to build and this specific kernel
flawlessly.
For readability during the RFC, I can remove them but without them, the
current kernel version won't build and run.

But I think the code for ODROID-U3 support has been merged upstream some
time ago and the next think I need to test is the mainline kernel. 

> Does it need xorg? can it not run on the fbdev or is there no hardware
> support?
>

It's not needed but the fbdev alternative have issues with vsync that
are not trivial to fix because fbdev support in Linux in not maintained
anymore and will be removed very soon in the mainline kernel.  And with
fbdev, there's no openegl also for fancy skins.

Thks !

-- 
Jérôme Benoit aka fraggle
Piment Noir - http://piment-noir.org
OpenPGP Key ID : 9FE9161D
Key fingerprint : 9CA4 0249 AF57 A35B 34B3 AC15 FAA0 CB50 9FE9 161D

File: packages/graphics/opengl-mali/package.mk [hunk 1]
```diff
@@ -17,13 +17,21 @@
...
+PKG_DEPENDS_TARGET="toolchain"
+if [ "$DISPLAYSERVER" = "x11" ]; then
+  PKG_VERSION="${PKG_VERSION}_x11"
+  PKG_URL="http://people.piment-noir.org/~fraggle/download/odroid/$PKG_NAME-$PKG_VERSION.tar.xz"
+  PKG_DEPENDS_TARGET="$PKG_DEPENDS_TARGET libXfixes libXrender libXcomposite libxcb"
+else
+  PKG_VERSION="${PKG_VERSION}_fbdev"
+  PKG_URL="http://people.piment-noir.org/~fraggle/download/odroid/$PKG_NAME-$PKG_VERSION.tar.xz"
+fi
+PKG_SOURCE_DIR="${PKG_NAME}-${PKG_VERSION}"
```
Comment (zalaare) [inline packages/graphics/opengl-mali/package.mk]: When this was written, I was constantly switching between fbdev and x11 flavors during testing.  If you plan to pick finalize your pick with _x11, then you can drop the _fbdev lines completely.

**Comment by chewitt** [comment]
@jerome-benoit LE welcomes new hardware, but would prefer to start with a community created and published build to establish the userbase before promoting the build project into our main git repo. This is partly because current team have no expertise with this hardware so we'd be forced to blindly review and accept any PR's to the project or branch. We're also being cautious as it's also the first time a user submits new hardware to us. We would be very happy to encourage and promote your efforts on the U2/U3 with a dedicated forum area to support users from, and if necessary webspace to help publishing. Does that sound like a reasonable compromise?

**Comment by jerome-benoit** [comment]
On Sun, 01 Jan 2017 11:35:55 -0800 in
<LibreELEC/LibreELEC.tv/pull/1077/c269916146@github.com>, Christian
Hewitt <notifications@github.com> wrote:

> @jerome-benoit LE welcomes new hardware, but would prefer to start
> with a community created and published build to establish the
> userbase before promoting the build project into our main git repo.
> This is partly because current team have no expertise with this
> hardware so we'd be forced to blindly review and accept any PR's to
> the project or branch. We're also being cautious as it's also the
> first time a user submits new hardware to us. We would be very happy
> to encourage and promote your efforts on the U2/U3 with a dedicated
> forum area to support users from, and if necessary webspace to help
> publishing. Does that sound like a reasonable compromise?

The user base of the ODROID-U2 LE portage have been promoted on the
ODROID forum :
http://forum.odroid.com/viewtopic.php?f=79&t=5915&hilit=zalaare (yes,
it's OE images but LE portage have been also done on github :
https://github.com/zalaare/Project-Odroid-U2). I do not know how many
people have used it. 
This PR is based on that work and include some enhancements and bug
fixes and is not mergeable yet.

It's fine to promote my work on LE forums but no need to create a new
section, the hardware section and a sticky topic is good enough.

And happy new year.

Cheers, 

-- 
Jérôme Benoit aka fraggle
Piment Noir - http://piment-noir.org
OpenPGP Key ID : 9FE9161D
Key fingerprint : 9CA4 0249 AF57 A35B 34B3 AC15 FAA0 CB50 9FE9 161D

**Comment by chewitt** [comment]
@jerome-benoit As long as the update function in the settings add-on still checks our servers we can periodically tell you exactly how many active installs there are. Currently 5x, but Rome wasn't built in a day :)

**Comment by jerome-benoit** [comment]
On Mon, 02 Jan 2017 21:22:41 -0800 in
<LibreELEC/LibreELEC.tv/pull/1077/c270053637@github.com>, Christian
Hewitt <notifications@github.com> wrote:

> @jerome-benoit As long as the update function in the settings add-on
> still checks our servers we can periodically tell you exactly how
> many active installs there are. 

Hum, it's default the configuration for now, which should check for
a non existant /Odroid_U2 URL for LE update and LE add-ons.  
I plan to change this to check for imx6 or RPi2 instead to permit
users to install LE add-ons (ABI compatible projects). 
If I do not change the update URL, users will have to do update
manually. If I do change the update URL, users will have auto update
feature working but I will then have no statistics.   
It's more or less the reason why I wanted use our current
infrastructure.

++

-- 
Jérôme Benoit aka fraggle
Piment Noir - http://piment-noir.org
OpenPGP Key ID : 9FE9161D
Key fingerprint : 9CA4 0249 AF57 A35B 34B3 AC15 FAA0 CB50 9FE9 161D

**Comment by chewitt** [comment]
@jerome-benoit I'm closing the PR down so team members stop receiving email updates as you push changes to your development branch. Thanks.

**Comment by jerome-benoit** [comment]
On Wed, 25 Jan 2017 21:43:42 -0800 in <LibreELEC/LibreELEC.tv/pull/1077/c275312333@github.com>, 
Christian Hewitt <notifications@github.com> wrote:

> @jerome-benoit I'm closing the PR down so team members stop receiving email updates as you push changes to your development branch. Thanks.

No pb.

-- 
Jérôme Benoit aka fraggle
Piment Noir - http://piment-noir.org
OpenPGP Key ID : 9FE9161D
Key fingerprint : 9CA4 0249 AF57 A35B 34B3 AC15 FAA0 CB50 9FE9 161D

**Comment by rzr** [comment]
Are those changes released ?
branch is no more public at
https://github.com/PimentNoir/LibreELEC.tv/tree/rfc-odroid-u2

**Comment by CvH** [comment]
@rzr those changes are not at LE and due the fact that PR is 3 years old it is likely already heavily outdated and not working anymore

**Comment by rzr** [comment]
OK thanks for info, I'll track related upstream changes ie:

https://github.com/hexdump0815/linux-mainline-and-mali-on-odroid-u3

**Pull request closed without merging**